sdks/java/io/hadoop-input-format/src/main/java/org/apache/beam/sdk/io/hadoop/inputformat/HadoopInputFormatIO.java

```diff
@@ -0,0 +1,1121 @@
+            inputSplit,
+            inputFormatObj,
+            taskAttemptContext);
+      }
+    }
+
+    /**
+     * BoundedReader for Hadoop InputFormat source.
+     *
+     * @param <K> Type of keys RecordReader emits.
+     * @param <V> Type of values RecordReader emits.
+     */
+    class HadoopInputFormatReader<T1, T2> extends BoundedSource.BoundedReader<KV<K, V>> {
+
+      private final HadoopInputFormatBoundedSource<K, V> source;
+      @Nullable private final SimpleFunction<T1, K> keyTranslationFunction;
+      @Nullable private final SimpleFunction<T2, V> valueTranslationFunction;
+      private final SerializableSplit split;
+      private RecordReader<T1, T2> currentReader;
+      private volatile boolean doneReading = false;
+      private long recordsReturned = 0L;
+      // Variable added to track the progress of the RecordReader.
+      private AtomicDouble progressValue = new AtomicDouble();
+      private transient InputFormat<T1, T2> inputFormatObj;
+      private transient TaskAttemptContext taskAttemptContext;
+
+      private HadoopInputFormatReader(HadoopInputFormatBoundedSource<K, V> source,
+          @Nullable SimpleFunction keyTranslationFunction,
+          @Nullable SimpleFunction valueTranslationFunction,
+          SerializableSplit split,
+          InputFormat inputFormatObj,
+          TaskAttemptContext taskAttemptContext) {
+        this.source = source;
+        this.keyTranslationFunction = keyTranslationFunction;
+        this.valueTranslationFunction = valueTranslationFunction;
+        this.split = split;
+        this.inputFormatObj = inputFormatObj;
+        this.taskAttemptContext = taskAttemptContext;
+      }
+
+      @Override
+      public HadoopInputFormatBoundedSource<K, V> getCurrentSource() {
+        return source;
+      }
+
+      @Override
+      public boolean start() throws IOException {
+        try {
+          recordsReturned = 0;
+          currentReader =
+              (RecordReader<T1, T2>) inputFormatObj.createRecordReader(split.getSplit(), taskAttemptContext);
+          if (currentReader != null) {
+            /*
+             * CurrentReader object could be accessed concurrently by multiple sources. Hence, to be
+             * on safer side, it has been added in synchronized block.
+             */
+            synchronized (currentReader) {
+              currentReader.initialize(split.getSplit(), taskAttemptContext);
+              if (currentReader.nextKeyValue()) {
+                recordsReturned++;
+                return true;
+              }
+            }
+          } else {
+            throw new IOException(String.format(
+                HadoopInputFormatIOConstants.NULL_CREATE_RECORDREADER_ERROR_MSG,
+                inputFormatObj.getClass()));
+          }
+          synchronized (currentReader) {
+            currentReader = null;
+          }
+        } catch (InterruptedException e) {
+          throw new IOException("Unable to read data: ", e);
```

Comment:
Done. Updated the comment as below:
throw new IOException(
              "Could not read because the thread got interrupted while reading the records with an exception: ",
              e);